File: clients/Java/3.4 mcp_delete_source/MCPDeleteSourceClient.java
import org.json.JSONObject;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.nio.charset.StandardCharsets;

public class MCPDeleteSourceClient {

    public static void main(String[] args) {
        // Minimaler Check, ob genügend Argumente übergeben wurden
        if (args.length < 4) {
            printUsage();
            return;
        }

        String serverIp    = getArgument(args, "--server-ip");
        String portStr     = getArgument(args, "--server-port");
        String token       = getArgument(args, "--token");
        String sourceId    = getArgument(args, "--source-id");

        // Wenn eines der erforderlichen Argumente null ist, Usage anzeigen
        if (serverIp == null || portStr == null || token == null || sourceId == null) {
            printUsage();
            return;
        }

        int serverPort = Integer.parseInt(portStr);

        System.out.println("📤 Sending request to delete source...");

        // JSON-Payload erstellen
        JSONObject payload = new JSONObject();
        payload.put("command", "delete_source");
        payload.put("token", token);

        JSONObject arguments = new JSONObject();
        arguments.put("sourceId", sourceId);

        payload.put("arguments", arguments);

        // Anfrage an den Server schicken
        String response = sendRequest(serverIp, serverPort, payload);
        System.out.println("✔️ Response from server:");
        System.out.println(response);
    }

    /**
     * Liest den Wert eines Arguments aus (z.B. --server-ip 127.0.0.1).
     * Gibt null zurück, wenn der Schlüssel nicht gefunden wurde.
     */
    private static String getArgument(String[] args, String key) {
        for (int i = 0; i < args.length - 1; i++) {
            if (args[i].equals(key)) {
                return args[i + 1];
            }
        }
        return null;
    }

    /**
     * Baut eine TCP-Verbindung zum Server auf, sendet das JSON-Payload
     * und empfängt die Antwort.
     */
    private static String sendRequest(String serverIp, int serverPort, JSONObject payload) {
        String payloadJson = payload.toString();

        try (Socket client = new Socket(serverIp, serverPort)) {
            // Payload senden
            OutputStream out = client.getOutputStream();
            byte[] data = payloadJson.getBytes(StandardCharsets.UTF_8);
            out.write(data);
            out.flush();

            // Antwort empfangen
            InputStream in = client.getInputStream();
            byte[] buffer = new byte[4096];
            StringBuilder responseBuilder = new StringBuilder();
            int bytesRead;

            do {
                bytesRead = in.read(buffer);
                if (bytesRead > 0) {
                    responseBuilder.append(new String(buffer, 0, bytesRead, StandardCharsets.UTF_8));
                }
            } while (bytesRead == buffer.length);

            return responseBuilder.toString();

        } catch (IOException e) {
            return "Error: " + e.getMessage();
        }
    }

    private static void printUsage() {
        System.out.println("Usage: --server-ip <IP> --server-port <PORT> "
                         + "--token <TOKEN> --source-id <SOURCE_ID>");
        System.out.println();
        System.out.println("Example:");
        System.out.println("  java -cp .;json-20241224.jar MCPDeleteSourceClient "
                         + "--server-ip 127.0.0.1 --server-port 1234 "
                         + "--token MyToken --source-id 12345");
    }
}

File: clients/Python/3.4 mcp_delete_source/MCPDeleteSourceClient.py
import socket
import json
import argparse

def delete_source(server_ip, server_port, token, source_id):
    """
    Sends a request to the MCP server to delete an existing source.

    :param server_ip: IP address of the MCP server
    :param server_port: Port number of the MCP server
    :param token: Authorization token
    :param source_id: ID of the source to delete
    :return: Response from the server
    """
    payload = {
        "command": "delete_source",
        "token": token,
        "arguments": {
            "sourceId": source_id
        }
    }

    payload_json = json.dumps(payload)

    try:
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as client_socket:
            client_socket.connect((server_ip, server_port))
            client_socket.sendall(payload_json.encode('utf-8'))

            response = b""
            while True:
                part = client_socket.recv(4096)
                response += part
                if len(part) < 4096:
                    break

            return response.decode('utf-8')
    except Exception as e:
        return f"Error: {e}"

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Delete a source from the MCP server.")
    parser.add_argument("--server-ip", required=True, help="IP address of the MCP server")
    parser.add_argument("--server-port", type=int, required=True, help="Port number of the MCP server")
    parser.add_argument("--token", required=True, help="Authorization token")
    parser.add_argument("--source-id", required=True, help="ID of the source to delete")

    args = parser.parse_args()

    response = delete_source(args.server_ip, args.server_port, args.token, args.source_id)
    print("Response from server:", response)
